crates/runtime/src/traphandlers.rs

```diff
@@ -23,40 +25,249 @@ extern "C" {
+                handler.sa_flags = libc::SA_SIGINFO | libc::SA_NODEFER | libc::SA_ONSTACK;
+                handler.sa_sigaction = trap_handler as usize;
+                libc::sigemptyset(&mut handler.sa_mask);
+                if libc::sigaction(signal, &handler, slot.as_mut_ptr()) != 0 {
+                    panic!(
+                        "unable to install SIGSEGV handler: {}",
+                        io::Error::last_os_error(),
+                    );
+                }
+            };
+
+            // Allow handling OOB with signals on all architectures
+            register(&mut PREV_SIGSEGV, libc::SIGSEGV);
+
+            // Handle `unreachable` instructions which execute `ud2` right now
+            register(&mut PREV_SIGILL, libc::SIGILL);
+
+            // x86 uses SIGFPE to report division by zero
+            if cfg!(target_arch = "x86") || cfg!(target_arch = "x86_64") {
+                register(&mut PREV_SIGFPE, libc::SIGFPE);
+            }
+
+            // On ARM, handle Unaligned Accesses.
+            // On Darwin, guard page accesses are raised as SIGBUS.
+            if cfg!(target_arch = "arm") || cfg!(target_os = "macos") {
+                register(&mut PREV_SIGBUS, libc::SIGBUS);
+            }
+        }
+
+        unsafe extern "C" fn trap_handler(
             signum: libc::c_int,
             siginfo: *mut libc::siginfo_t,
             context: *mut libc::c_void,
-        ) -> *const u8 {
-            tls::with(|info| {
-                match info {
-                    Some(info) => info.handle_trap(pc, false, |handler| handler(signum, siginfo, context)),
-                    None => ptr::null(),
+        ) {
+            let previous = match signum {
+                libc::SIGSEGV => &PREV_SIGSEGV,
+                libc::SIGBUS => &PREV_SIGBUS,
+                libc::SIGFPE => &PREV_SIGFPE,
+                libc::SIGILL => &PREV_SIGILL,
+                _ => panic!("unknown signal: {}", signum),
+            };
+            let handled = tls::with(|info| {
+                // If no wasm code is executing, we don't handle this as a wasm
+                // trap.
+                let info = match info {
+                    Some(info) => info,
+                    None => return false,
+                };
+
+                // If we hit an exception while handling a previous trap, that's
+                // quite bad, so bail out and let the system handle this
+                // recursive segfault.
+                //
+                // Otherwise flag ourselves as handling a trap, do the trap
+                // handling, and reset our trap handling flag. Then we figure
+                // out what to do based on the result of the trap handling.
+                let jmp_buf = info.handle_trap(
+                    get_pc(context),
+                    false,
+                    |handler| handler(signum, siginfo, context),
+                );
+                info.handling_trap.set(false);
```

Comment:
Why do we want to reset the handling_trap flag here? It seems that `handle_trap` already takes care about this and windows version doesn't do this either.

Reply:
Oops my mistake, forgot to move this when refactoring.